ssz_serialization/merkleization.nim

```diff
@@ -165,42 +177,46 @@ func computeZeroHashes: array[sizeof(Limit) * 8, Digest] =
 
 const zeroHashes* = computeZeroHashes()
 
-template combineChunks(start: int) =
-  for i in start .. merkleizer.topIndex:
+template combineChunks(merkleizer: var SszMerkleizerImpl, start: int) =
+  for i in start..<merkleizer.topIndex:
     trs "CALLING MERGE BRANCHES"
-    doAssert i != merkleizer.topIndex
     if getBitLE(merkleizer.totalChunks, i + 1):
-      merkleizer.combinedChunks[i + 1][1] = mergeBranches(
+      # chunkPtr avoids a temporary in the generated nim code
+      let chunkPtr = addr merkleizer.combinedChunks[i + 1][1]
+      chunkPtr[] = mergeBranches(
         merkleizer.combinedChunks[i][0], merkleizer.combinedChunks[i][1])
     else:
-      merkleizer.combinedChunks[i + 1][0] = mergeBranches(
+      let chunkPtr = addr merkleizer.combinedChunks[i + 1][0]
+      chunkPtr[] = mergeBranches(
         merkleizer.combinedChunks[i][0], merkleizer.combinedChunks[i][1])
       break
 
-func addChunk*(merkleizer: var SszMerkleizerImpl, data: openArray[byte]) =
-  doAssert data.len > 0 and data.len <= bytesPerChunk
-
+template addChunkDirect(merkleizer: var SszMerkleizerImpl, body: untyped) =
   if getBitLE(merkleizer.totalChunks, 0):
-    assign(merkleizer.combinedChunks[0][1].data.toOpenArray(0, data.high), data)
-    if data.len < bytesPerChunk:
-      zeroMem(
-        addr merkleizer.combinedChunks[0][1].data[data.len],
-        bytesPerChunk - data.len)
+    template chunk: Digest {.inject.} =
+      merkleizer.combinedChunks[0][1]
+    body
 
-    combineChunks(0)
+    merkleizer.combineChunks(0)
 
   else:
-    assign(merkleizer.combinedChunks[0][0].data.toOpenArray(0, data.high), data)
-    if data.len < bytesPerChunk:
-      zeroMem(
-        addr merkleizer.combinedChunks[0][0].data[data.len],
-        bytesPerChunk - data.len)
-
+    template chunk: Digest {.inject.} =
+      merkleizer.combinedChunks[0][0]
+    body
     trs "WROTE BASE CHUNK ",
-      toHex(merkleizer.combinedChunks[0].data), " ", data.len
+      toHex(merkleizer.combinedChunks[0][0].data), " ", data.len
 
   inc merkleizer.totalChunks
 
+
+
+func addChunk*(merkleizer: var SszMerkleizerImpl, data: openArray[byte]) =
+  doAssert data.len > 0 and data.len <= bytesPerChunk
+  addChunkDirect(merkleizer):
```

Comment:
What does `addChunkDirect` do semantically? The name is not very descriptive and as far as I can tell it does the usual merkle tree updates.

Reply:
it RVO:s the chunk straight into the cache instead of going via a temporary

Reply:
f334ba9ff748548e3ad12434f76f03d37b92c003

Reply:
in particular for fields, it allows writing the chunk hash for each field without generating 2 temporaries for every field (as the current code does)